mithril-client/src/cardano_database_client/download_unpack.rs

```diff
@@ -0,0 +1,1204 @@
+        for location in locations_sorted {
+            let immutable_files_numbers_downloaded = self
+                .download_unpack_immutable_files_for_location(
+                    &location,
+                    &immutable_file_numbers_to_download,
+                    compression_algorithm,
+                    immutable_files_target_dir,
+                    max_parallel_downloads,
+                    download_id,
+                )
+                .await?;
+            for immutable_file_number in immutable_files_numbers_downloaded {
+                immutable_file_numbers_to_download.remove(&immutable_file_number);
+            }
+            if immutable_file_numbers_to_download.is_empty() {
+                return Ok(());
+            }
+        }
+
+        Err(anyhow!(
+                "Failed downloading and unpacking immutable files for immutable_file_numbers: {immutable_file_numbers_to_download:?}"
+            ))
+    }
+
+    /// Download and unpack the immutable files of the given range.
+    ///
+    /// The download is attempted for each location until the full range is downloaded.
+    /// An error is returned if not all the files are downloaded.
+    async fn batch_download_unpack_immutable_files(
+        &self,
+        file_downloader: Arc<dyn FileDownloader>,
+        file_downloader_uris_chunk: Vec<(ImmutableFileNumber, FileDownloaderUri)>,
+        compression_algorithm: &CompressionAlgorithm,
+        immutable_files_target_dir: &Path,
+        download_id: &str,
+        file_size: u64,
+    ) -> MithrilResult<BTreeSet<ImmutableFileNumber>> {
+        let mut immutable_file_numbers_downloaded = BTreeSet::new();
+        let mut join_set: JoinSet<MithrilResult<ImmutableFileNumber>> = JoinSet::new();
+        for (immutable_file_number, file_downloader_uri) in file_downloader_uris_chunk.into_iter() {
+            join_set.spawn(self.spawn_immutable_download_future(
+                DownloadImmutableFutureBuilderArgs {
+                    file_downloader: file_downloader.clone(),
+                    immutable_file_number,
+                    file_downloader_uri,
+                    compression_algorithm: compression_algorithm.to_owned(),
+                    immutable_files_target_dir: immutable_files_target_dir.to_path_buf(),
+                    download_id: download_id.to_string(),
+                    file_size,
+                },
+            )?);
+        }
+        while let Some(result) = join_set.join_next().await {
+            match result? {
+                Ok(immutable_file_number) => {
+                    immutable_file_numbers_downloaded.insert(immutable_file_number);
+                }
+                Err(e) => {
+                    slog::error!(
+                        self.logger,
+                        "Failed downloading and unpacking immutable files"; "error" => e.to_string(), "target_dir" => immutable_files_target_dir.display()
```

Comment:
Should we debug the error to show the stack trace ? Else only the main message will be shown.
```suggestion
                        "Failed downloading and unpacking immutable files"; "error" => ?e, "target_dir" => immutable_files_target_dir.display()
```

Reply:
I modified the code to show the stack trace here and in the other places with the same problem.
Thanks for pointing that out :+1: